content/doc/book/pipeline/pipeline-best-practices.adoc

```diff
@@ -0,0 +1,97 @@
+---
+layout: section
+title: Pipeline Best Practices
+---
+ifdef::backend-html5[]
+:notitle:
+:description:
+:author: Alex Taylor
+:email: jenkinsci-docs@googlegroups.com
+:sectanchors:
+ifdef::env-github[:imagesdir: ../resources]
+ifndef::env-github[:imagesdir: ../../resources]
+:hide-uri-scheme:
+:toc:
+endif::[]
+
+= Pipeline Best Practices
+
+== General
+
+=== Groovy Code in Pipelines as Glue
+
+Make sure to use Groovy code as a way of wiring together a set of actions rather than the main functionality. 
+This means that rather than relying on Pipeline functionality(groovy or Pipeline steps) to drive the build process forward, it is better to use single steps(such as `sh`) to accomplish multiple parts of the build. 
+As the complexity of Pipelines increase (in terms of amount of Groovy code, number of steps used, etc.), they require more resources (CPU, memory, storage) on the master.
+Think of Pipeline as a tool to accomplish a build rather than it being the core of that build.
+
+Example: Using a single maven build step to drive the build through its build/test/deploy process.
+
+=== Avoid complex Groovy code in Pipeline
+
+Groovy code within a Pipeline uses master resources(memory and CPU). 
+For a Pipeline, groovy code *always* executes on master which is why it does use master resources. 
+Because of this, the most critical action item is to reduce the number of groovy calls that need to be made (this includes the many imports of groovy methods being used). 
+The most common example methods are things such as:
+. *JsonSlurper:* This function (and other similar ones like XmlSlurper or ReadFile) would be to read in some file which is on disk into a JSON object within the Pipeline so it would look something like this: `def JsonFile = new JsonSlurper().parseText(readFile("$LOCAL_FILE")))`. The issue here is that this loads that file into memory 2x on the master and if that file is very large or this command is being done very often then it will drain a lot of unnecessary memory
+.. Solution: Instead of using JsonSlurper, use a shell step and return the standard out. This shell would look something like this: `def JsonReturn = sh label: '', returnStdout: true, script: 'echo "$LOCAL_FILE"| jq "."'`. This will use agent resources to read the file and even parse the file with a good `jq` filter
```

Comment:
For me this is not really a "problem/solution" type of document but more to call out when people are using bad methods and how they should fix them. This doc is intended more to be once you are already using pipelines extensively and have already set most things up rather than a `how to`